File: MicrocodeCompiler/AST.cs
using System;
using System.Collections.Generic;

namespace MicrocodeCompiler.AST {
    class Microcode {
        public readonly Header header;
        public readonly List<Opcode> opcodes;

        public Microcode(Header header, List<Opcode> opcodes) {
            this.header = header;
            this.opcodes = opcodes;
        }

        public void Print() {
            Console.Write($"Microcode {{ ");
            if(header != null)header.Print();
            Console.Write(", ");
            foreach(var line in opcodes) {
                line.Print();
            }
            Console.Write(" }");
        }
    }

    class Block { }

    class Header : Block {
        public readonly List<Line> lines;

        public Header(List<Line> lines) {
            this.lines = lines;
        }

        public void Print() {
            Console.Write("Header { ");
            foreach(var line in lines) {
                line.Print();
            }
            Console.Write(" }");
        }
    }
    
    class Opcode : Block {
        public readonly string name;
        public readonly List<string> arguments;
        public readonly int code;
        public readonly List<Line> lines;

        public Opcode(string name, List<string> arguments, int code, List<Line> lines) {
            this.name = name;
            this.arguments = arguments;
            this.code = code;
            this.lines = lines;
        }

        public void Print() {
            Console.Write($"Opcode({name}) = {code} {{ ");
            foreach(var line in lines) {
                line.Print();
            }
            Console.Write(" }");
        }
    }
    class Line {
        public readonly Condtion run;
        public readonly List<string> bits;

        public Line(Condtion run, List<string> bits) {
            this.run = run;
            this.bits = bits;
        }

        public void Print() {
            Console.Write("Line { ");
            run.Print();
            foreach(var item in bits) {
                Console.Write($", {item}");
            }
            Console.Write(" } ");
        }
    }
    class Condtion {
        public readonly bool AlwaysRun;
        public readonly string condition;
        public readonly int conditionValue;

        public Condtion() {
            AlwaysRun = true;
        }

        public Condtion(string condition, int conditionValue) {
            AlwaysRun = false;
            this.condition = condition;
            this.conditionValue = conditionValue;
        }

        public void Print() {
            if(AlwaysRun) {
                Console.Write("Condition { Always }");
            } else {
                Console.Write($"Condition {{ {condition} = {conditionValue} }}");
            }
        }
    }
}

File: microasm/src/ast.c
#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>
#include "token.h"
#include "ast.h"
#include "error.h"
#include "token.h"
#include "memory.h"

void InitMicrocode(Microcode* mcode, const char* fileName) {
    mcode->hasError = false;
    mcode->fileName = fileName;
    ARRAY_ZERO(mcode->head, bit);
    ARRAY_ZERO(mcode->inp, value);
    ARRAY_ZERO(mcode->out, value);
    ARRAY_ALLOC(OpCode, *mcode, opcode);
    ARRAY_ALLOC(Error, *mcode, error);
#ifdef debug
    ARRAY_ALLOC(Error, *mcode, expectedError);
#endif
}

void PrintMicrocode(Microcode* mcode) {
    printf("Microcode:\n");
    printf("  Has Error: %s\n", mcode->hasError ? "true" : "false");
    printf("  File Name: %s\n", mcode->fileName);
    printf("  Header: %u", mcode->head.bitCount);
    for(unsigned int i = 0; i < mcode->head.bitCount; i++) {
        printf("\n    ");
        TokenPrint(&mcode->head.bits[i]);
    }
    printf("\n");
    printf("  Input: %u", mcode->inp.valueCount);
    for(unsigned int i = 0; i < mcode->inp.valueCount; i++) {
        printf("\n    ");
        TokenPrint(&mcode->inp.values[i].name);
        printf(" = %i", mcode->inp.values[i].value.data.value);
    }
    printf("\n");
    printf("  Output(%u): %u", mcode->out.width.data.value, mcode->out.valueCount);
    for(unsigned int i = 0; i < mcode->out.valueCount; i++) {
        printf("\n    ");
        printf("%i = ", mcode->out.values[i].id.data.value);
        TokenPrint(&mcode->out.values[i].name);
    }
    printf("\n");
    printf("  OpCodes: %u", mcode->opcodeCount);
    for(unsigned int i = 0; i < mcode->opcodeCount; i++) {
        OpCode* code = &mcode->opcodes[i];
        printf("\n    OpCode: ");
        TokenPrint(&code->name);
        printf(" = %u", code->id.data.value);
        printf("\n      Parameters: %u", code->parameterCount);
        for(unsigned int j = 0; j < code->parameterCount; j++) {
            printf("\n        ");
            TokenPrint(&code->parameters[j]);
        }
        printf("\n      Lines: %u", code->lineCount);
        for(unsigned int j = 0; j < code->lineCount; j++) {
            Line* line = code->lines[j];
            printf("\n        Line %u:", j);
            if(line->anyCondition) {
                printf("\n          Conditions: any");
            } else {
                printf("\n          Conditions: %u", line->conditionCount);
                for(unsigned int k = 0; k < line->conditionCount; k++) {
                    printf("\n            ");
                    TokenPrint(&line->conditions[k].name);
                    printf(" = %u", line->conditions[k].value.data.value);
                }
            }
            printf("\n          Bits: %u", line->bitCount);
            for(unsigned int k = 0; k < line->bitCount; k++) {
                printf("\n            ");
                TokenPrint(&line->bits[k]);
            }
        }
    }
    printf("\n");
}